src/Repository/ProfileField.php

```diff
@@ -0,0 +1,296 @@
+			$profileFieldOrder[$newProfileField->id] = $profileFieldOrder['new'];
+
+			$profileFields[] = $newProfileField;
+		}
+
+		unset($profileFieldInputs['new']);
+		unset($profileFieldOrder['new']);
+
+		// Prunes profile field whose label has been emptied
+		$profileFields = $profileFields->filter(function (Model\ProfileField $profileField) use (&$profileFieldInputs, &$profileFieldOrder) {
+			$keepModel = !isset($profileFieldInputs[$profileField->id]) || !empty($profileFieldInputs[$profileField->id]['label']);
+
+			if (!$keepModel) {
+				unset($profileFieldInputs[$profileField->id]);
+				unset($profileFieldOrder[$profileField->id]);
+				$this->delete($profileField);
+			}
+
+			return $keepModel;
+		});
+
+		// Regenerates the order values if items were deleted
+		$profileFieldOrder = array_flip(array_keys($profileFieldOrder));
+
+		// Update existing profile fields from form values
+		$profileFields = $profileFields->map(function (Model\ProfileField $profileField) use ($uid, $aclFormatter, &$profileFieldInputs, &$profileFieldOrder) {
+			if (isset($profileFieldInputs[$profileField->id]) && isset($profileFieldOrder[$profileField->id])) {
+				$psid = $this->permissionSet->getIdFromACL(
+					$uid,
+					$aclFormatter->toString($profileFieldInputs[$profileField->id]['contact_allow'] ?? ''),
+					$aclFormatter->toString($profileFieldInputs[$profileField->id]['group_allow'] ?? ''),
+					$aclFormatter->toString($profileFieldInputs[$profileField->id]['contact_deny'] ?? ''),
+					$aclFormatter->toString($profileFieldInputs[$profileField->id]['group_deny'] ?? '')
+				);
+
+				$profileField->psid  = $psid;
+				$profileField->label = $profileFieldInputs[$profileField->id]['label'];
+				$profileField->value = $profileFieldInputs[$profileField->id]['value'];
+				$profileField->order = $profileFieldOrder[$profileField->id];
+
+				unset($profileFieldInputs[$profileField->id]);
+				unset($profileFieldOrder[$profileField->id]);
+			}
+
+			return $profileField;
+		});
+
+		return $profileFields;
+	}
+
+	/**
+	 * Migrates a legacy profile to the new slimmer profile with extra custom fields.
+	 * Multi profiles are converted to ACl-protected custom fields and deleted.
+	 *
+	 * @param array $profile Profile table row
+	 * @throws \Exception
+	 */
+	public function migrateFromProfile(array $profile)
+	{
+		// Already processed, aborting
+		if ($profile['is-default'] === null) {
+			return;
+		}
+
+		if (!$profile['is-default']) {
+			$contacts = Model\Contact::selectToArray(['id'], ['uid' => $profile['uid'], 'profile-id' => $profile['id']]);
+			if (!count($contacts)) {
+				// No contact visibility selected defaults to user-only permission
+				$contacts = Model\Contact::selectToArray(['id'], ['uid' => $profile['uid'], 'self' => true]);
+			}
+
+			$allow_cid = DI::aclFormatter()->toString(array_column($contacts, 'id'));
+		}
+
+		$psid = DI::permissionSet()->getIdFromACL($profile['uid'], $allow_cid ?? '');
```

Comment:
change to `$this->repoPermissionSet`? Or is it the model call? I've to look, but nevertheless plz replace it with the right injected class :)

Reply:
There aren't any model calls that come from `DI` anymore with DDD.

Reply:
of course ... ;-)